arrays/arrays.c

```diff
@@ -69,43 +83,65 @@ void resize_array(Array *arr) {
  *
  * Throw an error if the index is out of range.
  *****/
+// fprintf writes formatted text to the output stream you specify.
+// printf is equivalent to writing fprintf(stdout, ...) and
+// writes formatted text to wherever the standard output stream is currently pointing.
+// sprintf writes formatted text to an array of char, as opposed to a stream.
 char *arr_read(Array *arr, int index) {
-
   // Throw an error if the index is greater than the current count
+  if (index > arr->count) {
+    fprintf(stderr, "Index is out of range.");
+    exit(1);
+  }
 
   // Otherwise, return the element at the given index
+  return arr->elements[index];
 }
 
 
 /*****
  * Insert an element to the array at the given index
  *****/
 void arr_insert(Array *arr, char *element, int index) {
-
   // Throw an error if the index is greater than the current count
+  if (index > arr->count) {
+    fprintf(stderr, "Index is out of range.");
+    exit(1);
+  }
 
-  // Resize the array if the number of elements is over capacity
+  // Resize the array if the number of elements is equal to or over capacity
+  if (arr->count >= arr->capacity) {
+    resize_array(arr);
+  }
 
   // Move every element after the insert index to the right one position
+  for (int i=arr->count; i > index; i--) {
+    arr->elements[i] = arr->elements[i-1]; // repointing the new address to previous element's address
+  }
 
   // Copy the element and add it to the array
+  char *new_element = element; // copy the element so it doesn't get lost
```

Comment:
Should be copying with strdup() in order to malloc() space for the element, so it would be har *new_element = strdup(element). It's not going to break your code if you don't have this though.

Reply:
Okay thanks!